### PR TITLE
Tsp and Mosaic

### PR DESCRIPTION
**Pull request by Preetesh21**
This PR is made for issue #1 
I have added the two files Mosaic and TSP_art where the Mosaic takes  in two inputs (one being the target image and the other being the images directory) and TSP_art takes in only one input the image. I have also updated the Readme accordingly. 

The TSP program takes a considerable amount of time depending upon the size of the graph(i.e. how many cities to be considered for TSP problem). I have used an optimization library for the same.
I have added the Tqdm for Mosaic but could not for TSP since I used a ready made function.

I would really like to thank you for motivating me to complete the task and would certainly love to work on other issues as well.
Thank you !!!

### CODE REVIEW

**Comment by smaranjitghose** [comment]
> This PR is made for issue #1
> I have added the two files Mosaic and TSP_art where the Mosaic takes in two inputs (one being the target image and the other being the images directory) and TSP_art takes in only one input the image. I have also updated the Readme accordingly.
> 
> The TSP program takes a considerable amount of time depending upon the size of the graph(i.e. how many cities to be considered for TSP problem). I have used an optimization library for the same.
> I have added the Tqdm for Mosaic but could not for TSP since I used a ready made function.
> 
> I would really like to thank you for motivating me to complete the task and would certainly love to work on other issues as well.
> Thank you !!!

See I told you with some patience and hustle, you can definitely do it. 😉
Don't thank me it's your effort and dedication.

@Preetesh21 can you please put the additional images used for generating the mosaic art inside ``assets/mosaic_images``. It would be really nice if you credit the sources of the images as well

File: tsp.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# pip install tsp_solver
```
Comment (smaranjitghose) [inline tsp.py]: Remove this comment  please

File: tsp.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# kindly note the code takes considerable amount of time to run depending upon the "size" parameter.
+import os  
+import cv2  
+from PIL import Image  
+import cv2
+from scipy.spatial.distance import pdist, squareform  
+from tsp_solver.greedy_numpy import solve_tsp
```
Comment (smaranjitghose) [inline tsp.py:7]: Can you elaborate from where are you importing this? Is it a python package or you wrote a script.

File: tsp.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ap.add_argument("-i", "--image", required=True, help="path to input image")
+args = vars(ap.parse_args())
+
+image_path = args['image'] 
+img=cv2.imread(image_path)
+img=cv2.resize(img, (580, 740),interpolation = cv2.INTER_AREA) 
+grayImage = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+(thresh, blackAndWhiteImage) = cv2.threshold(grayImage, 100, 255, cv2.THRESH_BINARY)  
+
+import numpy as np
```
Comment (smaranjitghose) [inline tsp.py]: Try to put all the import statements at the beginning of the script

File: tsp.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  
+bw_image_array = np.array(blackAndWhiteImage, dtype=np.int)  
+black_indices = np.argwhere(bw_image_array == 0)  
+# Changing "size" to a larger value makes this algorithm take longer,  
+# but provides more granularity to the portrait  
+chosen_black_indices = black_indices[np.random.choice(black_indices.shape[0],size=8000)]  
+
+distances = pdist(chosen_black_indices)  
+distance_matrix = squareform(distances)
+
```
Comment (smaranjitghose) [inline tsp.py:34]: Try to explain in bullet points (single-line comments about the how the tsp is working)

File: tsp.py (hunk 1 above)
Comment (Preetesh21) [reply tsp.py]: Done sir.

File: tsp.py (hunk 2 above)
Comment (Preetesh21) [reply tsp.py:7]: It is a Python Package, sir. Here is the [ link](https://pypi.org/project/tsp_solver/)

File: tsp.py (hunk 3 above)
Comment (Preetesh21) [reply tsp.py]: Done sir.

File: tsp.py (hunk 4 above)
Comment (Preetesh21) [reply tsp.py:34]: Done sir.

**Comment by Preetesh21** [comment]
Sir I had downloaded the images from Google Images. So where should I mention this to give proper reference

File: tsp.py (hunk 2 above)
Comment (Preetesh21) [reply tsp.py:7]: The Github repo [link](https://github.com/dmishin/tsp-solver)

**Comment by smaranjitghose** [comment]
> Sir I had downloaded the images from Google Images. So where should I mention this to give proper reference

Below the table

**Comment by Preetesh21** [comment]
> > Sir I had downloaded the images from Google Images. So where should I mention this to give proper reference
> 
> Below the table

Sure sir. Done.

File: tsp.py (hunk 2 above)
Comment (smaranjitghose) [reply tsp.py:7]: Okay!

**Review by smaranjitghose — APPROVED** [review]

**Comment by smaranjitghose** [comment]
@Preetesh21 Nice work man!

# One small suggestion:
- We wish to make this into a web app and if possible a python pacakge
- So we want it to be as lightweight as possible.
- Can you either zip the files used for the mosaic and push it inside the images folder ( delete the original ones then)
- Additionally, write a python script that fetches the images from the URLs ( store the URLs in a list)

**Pull request merged**